crates/bindings/README.md

```diff
@@ -0,0 +1,517 @@
+- `[your_project_directory]/target/doc/spacetimedb/struct.Local.html` (non-Windows)
+- `[your_project_directory]\target\doc\spacetimedb\struct.Local.html` (Windows)
+-->
+
+#### The `log` crate
+
+SpacetimeDB Rust modules have built-in support for the [log crate](log). All modules automatically install a suitable logger when they are first loaded by SpacetimeDB. (At time of writing, this happens [here](https://github.com/clockworklabs/SpacetimeDB/blob/e9e287b8aab638ba6e8bf9c5d41d632db041029c/crates/bindings/src/logger.rs)). Log macros can be used anywhere in module code, and log outputs of a running module can be inspected using the `spacetime logs` command:
+
+```text
+spacetime logs <DATABASE_IDENTITY>
+```
+
+<!-- TODO: we seem to have removed `println`, is that still hidden somewhere? -->
+
+#### Lifecycle Reducers
+
+A small group of reducers are called at set points in the module lifecycle. These are used to initialize
+the database and respond to client connections. See [Lifecycle Reducers](macro@crate::reducer#lifecycle-reducers).
+
+#### Scheduled Reducers
+
+Reducers can be scheduled to run repeatedly. This can be used to implement timers, game loops, and
+maintenance tasks. See [Scheduled Reducers](macro@crate::reducer#scheduled-reducers).
+
+## Automatic migrations
+
+When you `spacetime publish` a module that has already been published using `spacetime publish <DATABASE_IDENTITY>`,
+SpacetimeDB attempts to automatically migrate your existing database to the new schema. (The "schema" is just the collection
+of tables and reducers you've declared in your code, together with the types they depend on.) This form of migration is very limited and only supports a few kinds of changes.
+On the plus side, automatic migrations usually don't break clients. The situations that may break clients are documented below.
+
+The following changes are always allowed and never breaking:
+
+<!-- TODO: everything here should be smoke-tested. -->
+
+- ✅ **Adding tables**. Non-updated clients will not be able to see the new tables.
+- ✅ **Adding indexes**.
+- ✅ **Adding or removing `#[autoinc]` annotations.**
+- ✅ **Changing tables from private to public**.
+- ✅ **Adding reducers**.
+- ✅ **Removing `#[unique]`  annotations.**
+
+The following changes are allowed, but may break clients:
+
+- ⚠️ **Changing or removing reducers**. Clients that attempt to call the old version of a changed reducer will receive runtime errors.
+- ⚠️ **Changing tables from public to private**. Clients that are subscribed to a newly-private table will receive runtime errors.
+- ⚠️ **Removing `#[primary_key]` annotations**. Non-updated clients will still use the old `#[primary_key]` as a unique key in their local cache, which can result in non-deterministic behavior when updates are received.
```

Comment:
I don't think this breaks clients. It's certainly not non-deterministic. It would be a breaking change to downgrade `primary_key` to not-unique-at-all, but I think we only allow `primary_key` -> `unique`, which doesn't actually change client behavior at all.

Reply:
Removing `#[unique]` is allowed, so I don't see any logic that would catch what you're saying. 

...And the `primary_key` field is ignored by `auto_migrate_table`, whoops! https://github.com/clockworklabs/SpacetimeDB/issues/2262